index.js

```diff
@@ -65,24 +65,31 @@ const bindConnection = (options, aedesHandler, conn, req = {}) => {
   if (options.trustProxy) {
     extractConnectionDetails(options, aedesHandler, conn, req)
   } else {
-    const protocol = options.extractSocketDetails(conn.socket || conn)
-    req.connDetails = protocol
+    req.connDetails = options.extractSocketDetails(conn.socket || conn)
     aedesHandler(conn, req)
   }
 }
 
 const extractConnectionDetails = (options, aedesHandler, conn, req = {}) => {
-  const onData = (buffer) => {
-    const protocol = options.protocolDecoder(conn, buffer, req)
-    req.connDetails = protocol
-    conn.removeListener('data', onData)
-    conn.pause()
-    conn.unshift(protocol.data || buffer)
-    aedesHandler(conn, req)
+  // todo check drain event ?
+  const onReadable = (err) => {
+    if (err) {
+      return
+    }
+    const buffer = conn.read(null)
+    if (buffer) {
+      const protocol = options.protocolDecoder(conn, buffer, req)
```

Comment:
What if the buffer does not allow the full protocol to be parsed? this should be supported.

Reply:
This can happen, i think it depends on the proxy configuration. Still how can we guess beforehand that in the buffer a Proxy header is present ? Knowing that v1 of Proxy protocol header has not the same size as v2.
The trick would be to set an option that would declare which proxy version is used, so we can fill the buffer with the right amount of data, but it's not really flexible. 
I would consider that it's the responsibility of the proxy maintainer to ensure that the proxy is fully contained in one buffer.
What would be your approach ?

Reply:
My bad, the proxy protocol header cannot be broken down into multiple segments (ref. https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt) - can you add this link in a comment?

Reply:
Sure, done.